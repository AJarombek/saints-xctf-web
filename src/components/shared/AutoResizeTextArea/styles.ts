/**
 * JSS styles for the AutoResizeTextArea component.
 * @author Andrew Jarombek
 * @since 8/30/2020
 */

import { FontMixins } from '../../../styles/mixins';

export default {
  textArea: {
    padding: '6px 4px',
    height: '35px',
    backgroundColor: '#fdfdfd',
    border: '2px solid #bbb',
    borderRadius: '5px',
    ...FontMixins.roboto(),
    fontSize: '14px',
    resize: 'none',

    '&:focus': {
      outline: 'none',
    },
  },
};
