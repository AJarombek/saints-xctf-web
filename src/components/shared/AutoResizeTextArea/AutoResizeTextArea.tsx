/**
 * Component for a text area that grows as text is added and shrinks as text is removed.
 * @author Andrew Jarombek
 * @since 8/30/2020
 */

import React, { forwardRef, RefObject, useCallback, useEffect } from 'react';
import { createUseStyles } from 'react-jss';
import styles from './styles';
import classNames from 'classnames';
import { ClassValue } from 'classnames/types';

interface Props {
  maxLength: number;
  placeholder: string;
  onChange: (e: React.ChangeEvent<HTMLTextAreaElement>) => void;
  useCustomValue?: boolean;
  value?: string;
  disabled: boolean;
  className: ClassValue;
}

type TRef = HTMLTextAreaElement;

const useStyles = createUseStyles(styles);

// eslint-disable-next-line react/display-name
const AutoResizeTextArea = forwardRef<TRef, Props>(
  ({ maxLength, placeholder, onChange, useCustomValue, value, disabled, className }, ref: RefObject<any>) => {
    const classes = useStyles();

    const onKeyUp = useCallback((): void => {
      if (ref?.current) {
        ref.current.style.height = '25px';
        ref.current.style.height = `${ref.current.scrollHeight + 4}px`;
      }
    }, [ref]);

    useEffect(() => {
      if (ref) {
        onKeyUp();
      }
    }, [ref, onKeyUp]);

    return (
      <textarea
        className={classNames(classes.textArea, className)}
        maxLength={maxLength}
        placeholder={placeholder}
        value={useCustomValue ? value : ref?.current?.value}
        ref={ref}
        onChange={onChange}
        onKeyUp={onKeyUp}
        disabled={disabled}
      />
    );
  },
);

export default AutoResizeTextArea;
