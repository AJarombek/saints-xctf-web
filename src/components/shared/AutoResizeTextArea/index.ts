import AutoResizeTextArea from './AutoResizeTextArea';

export default AutoResizeTextArea;
