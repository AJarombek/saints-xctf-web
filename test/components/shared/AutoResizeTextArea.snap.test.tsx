/**
 * Snapshot test for the AutoResizeTextArea component.
 * @author Andrew Jarombek
 * @since 3/7/2021
 */

import React from 'react';
import renderer from 'react-test-renderer';
import AutoResizeTextArea from '../../../src/components/shared/AutoResizeTextArea';

describe('AutoResizeTextArea Snapshot Tests', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <AutoResizeTextArea
          maxLength={100}
          placeholder="Snapshot Text Area"
          onChange={(): void => null}
          disabled={false}
          className={null}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
