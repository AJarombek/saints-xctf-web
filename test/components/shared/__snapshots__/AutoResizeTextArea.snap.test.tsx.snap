// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoResizeTextArea Snapshot Tests renders correctly 1`] = `
<textarea
  className="textArea-0-1-1"
  disabled={false}
  maxLength={100}
  onChange={[Function]}
  onKeyUp={[Function]}
  placeholder="Snapshot Text Area"
/>
`;
